packages/my-queries-storage/src/hooks.ts

```diff
@@ -0,0 +1,76 @@
+import { useEffect } from 'react';
+import { useSyncExternalStore } from 'use-sync-external-store/shim';
+import type { FavoriteQuerySchema, RecentQuerySchema } from './query-storage';
+import { FavoriteQueryStorage, RecentQueryStorage } from './query-storage';
+
+let lastId = 0;
+const subscriptions = new Map<number, () => void>();
+const subscribe = (callback: () => void) => {
+  const subscriptionId = lastId++;
+  subscriptions.set(subscriptionId, callback);
+
+  return () => subscriptions.delete(subscriptionId);
+};
+
+const favoriteStorage = new FavoriteQueryStorage();
+const recentsStorage = new RecentQueryStorage();
+
+const state = {
+  favoriteQueries: [] as Array<typeof FavoriteQuerySchema>,
+  recentQueries: [] as Array<typeof RecentQuerySchema>,
+  loaded: false,
+};
+
+const setFavoriteQueries = (value: Array<typeof FavoriteQuerySchema>) =>
+  (state.favoriteQueries = value as any);
+const setRecentQueries = (value: Array<typeof RecentQuerySchema>) =>
+  (state.recentQueries = value as any);
+
+const notify = () => {
+  for (const cb of subscriptions.values()) {
+    cb();
+  }
+};
+
+export function useSavedQueries() {
+  if (!state.loaded) {
+    state.loaded = true;
+
+    void favoriteStorage
+      .loadAll()
+      .then((q) => setFavoriteQueries(q as any))
+      .then(notify);
```

Comment:
This shlud probably do some error handling and return it as part of the possible loading state?

Reply:
I totally agree. It might make sense to have some field in the state returned by the hook that exposes the error in a friendly manner to the component, so it can show a proper error if necessary.